pyboy/api/memoryscanner.py

```diff
@@ -0,0 +1,119 @@
+from enum import Enum
+
+from pyboy.utils import bcd_to_dec
+
+
+class StandardComparisonType(Enum):
+    """Enumeration for defining types of comparisons that do not require a previous value."""
+    EXACT = 1
+    LESS_THAN = 2
+    GREATER_THAN = 3
+    LESS_THAN_OR_EQUAL = 4
+    GREATER_THAN_OR_EQUAL = 5
+
+
+class DynamicComparisonType(Enum):
+    """Enumeration for defining types of comparisons that require a previous value."""
+    UNCHANGED = 1
+    CHANGED = 2
+    INCREASED = 3
+    DECREASED = 4
+    MATCH = 5
+
+
+class ScanMode(Enum):
+    """Enumeration for defining scanning modes."""
+    INT = 1
+    BCD = 2
+
+
+class MemoryScanner():
+    """A class for scanning memory within a given range."""
+    def __init__(self, pyboy):
+        """
+        Initializes the MemoryScanner with a pyboy instance.
+
+        :param pyboy: The pyboy emulator instance.
+        """
+        self.pyboy = pyboy
+        self._memory_cache = {}
+        self._memory_cache_byte_width = 1
+
+    def scan_memory(
+        self,
+        target_value=None,
+        start_addr=0x0000,
+        end_addr=0xFFFF,
+        standard_comparison_type=StandardComparisonType.EXACT,
+        value_type=ScanMode.INT,
+        byte_width=1,
+        byteorder="little"
+    ):
+        """
+        Scans memory in the specified range, looking for a target value.
+
+        :param start_addr: The starting address for the scan.
+        :param end_addr: The ending address for the scan.
+        :param target_value: The value to search for or None for any value (used in relation to `MemoryScanner.rescan_memory`).
+        :param standard_comparison_type: The type of comparison to use.
+        :param value_type: The type of value (INT or BCD) to consider.
+        :param byte_width: The number of bytes to consider for each value.
+        :param byteorder: The endian type to use (see [int.from_bytes](https://docs.python.org/3/library/stdtypes.html#int.from_bytes)). Note, this is only used for 16-bit values and higher.
+        :return: A list of addresses where the target value is found.
```

Comment:
The docs generator we're using doesn't pick up this format. You can see `pyboy.py` for examples on how to format it.